rset/rsets/helper.go

```
@@ -34,3 +36,145 @@ func HasAggFields(fields []*field.Field) bool {
 	aggFields := GetAggFields(fields)
 	return len(aggFields) > 0
 }
+
+// castIdent returns an Ident expression if e is or nil.
+func castIdent(e expression.Expression) *expression.Ident {
+	i, ok := e.(*expression.Ident)
+	if !ok {
+		return nil
+	}
+	return i
+}
+
+// TODO: export clause type and move to plan?
+type clauseType int
+
+const (
+	noneClause clauseType = iota
+	onClause
+	whereClause
+	groupByClause
+	fieldListClause
+	havingClause
+	orderByClause
+)
+
+func (clause clauseType) String() string {
+	switch clause {
+	case onClause:
+		return "on clause"
+	case fieldListClause:
+		return "field list"
+	case whereClause:
+		return "where clause"
+	case groupByClause:
+		return "group statement"
+	case orderByClause:
+		return "order clause"
+	case havingClause:
+		return "having clause"
+	}
+	return "none"
+}
+
+// castPosition returns an group/order by Position expression if e is a number.
+func castPosition(e expression.Expression, selectList *plans.SelectList, clause clauseType) (*expression.Position, error) {
+	v, ok := e.(expression.Value)
+	if !ok {
+		return nil, nil
```

Any need to return an error if expression is not expression.Value？